users/postgres/groups.go

```diff
@@ -0,0 +1,497 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"regexp"
+
+	"github.com/gofrs/uuid"
+	"github.com/lib/pq"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/users"
+)
+
+var (
+	errDeleteGroupDB = errors.New("delete group failed")
+	errSelectDb      = errors.New("select thing from db error")
+
+	errFK         = "foreign_key_violation"
+	errInvalid    = "invalid_text_representation"
+	errTruncation = "string_data_right_truncation"
+
+	groupRegexp = regexp.MustCompile("^[a-zA-Z0-9]+$")
+)
+
+var _ users.GroupRepository = (*groupRepository)(nil)
+
+type groupRepository struct {
+	db Database
+}
+
+// NewGroupRepo instantiates a PostgreSQL implementation of group
+// repository.
+func NewGroupRepo(db Database) users.GroupRepository {
+	return &groupRepository{
+		db: db,
+	}
+}
+
+func (gr groupRepository) Save(ctx context.Context, group users.Group) (users.Group, error) {
+	var id, q string
+	if group.Parent == nil || group.Parent.ID == "" {
+		q = `INSERT INTO groups (name, description, id, owner_id, metadata) VALUES (:name, :description, :id, :owner_id, :metadata) RETURNING id`
+	} else {
+		q = `INSERT INTO groups (name, description, id, owner_id, parent_id, metadata) VALUES (:name, :description, :id, :owner_id, :parent_id, :metadata) RETURNING id`
+	}
+
+	if group.ID == "" || group.Name == "" || !groupRegexp.MatchString(group.Name) {
+		return users.Group{}, users.ErrMalformedEntity
+	}
+
+	dbu, err := toDBGroup(group)
+	if err != nil {
+		return users.Group{}, err
+	}
+
+	row, err := gr.db.NamedQueryContext(ctx, q, dbu)
+	if err != nil {
+		pqErr, ok := err.(*pq.Error)
+		if ok {
+			switch pqErr.Code.Name() {
+			case errInvalid, errTruncation:
+				return users.Group{}, errors.Wrap(users.ErrMalformedEntity, err)
+			case errDuplicate:
+				return users.Group{}, errors.Wrap(users.ErrGroupConflict, err)
+			}
+		}
+
+		return users.Group{}, errors.Wrap(users.ErrCreateGroup, err)
+	}
+
+	defer row.Close()
+	row.Next()
+	if err := row.Scan(&id); err != nil {
+		return users.Group{}, err
+	}
+	group.ID = id
+	return group, nil
+}
+
+func (gr groupRepository) Update(ctx context.Context, group users.Group) error {
+	q := `UPDATE groups SET(name, description, metadata) VALUES (:name, :description, :metadata) WHERE id = :id`
+
+	dbu, err := toDBGroup(group)
+	if err != nil {
+		return errors.Wrap(errUpdateDB, err)
+	}
+	if _, err := gr.db.NamedExecContext(ctx, q, dbu); err != nil {
+		return errors.Wrap(errUpdateDB, err)
+	}
+
+	return nil
+}
+
+func (gr groupRepository) Delete(ctx context.Context, groupID string) error {
+	tx, err := gr.db.BeginTxx(context.Background(), nil)
```

Comment:
Looks like this check introduces a lot of complexity only to check if a group is empty. Leaving this check to the user and removing it from the code would simplify it. If there is a need, we can add it later.